clang/lib/AST/ASTDumper.cpp

```diff
@@ -357,6 +357,38 @@ void ASTDumper::VisitCompoundStmt(const CompoundStmt *Node) {
   }
 }
 
+void ASTDumper::VisitWhereClause(const WhereClause *WC) {
+  if (!WC || !DumpWhereClauses)
+    return;
+
+  NodeDumper.AddChild([=] {
+    ColorScope Color(OS, ShowColors, StmtColor);
+    OS << "WhereClause";
+    
+    FactListTy Facts = WC->getFacts(); 
+    for (auto *Fact : Facts) {
+      VisitWhereClauseFact(Fact);
+      }
+  }); 
+}
+
+void ASTDumper::VisitWhereClauseFact(const WhereClauseFact *Fact) {
+  if (auto *BF = dyn_cast<BoundsDeclFact>(Fact)) {    
+    NodeDumper.AddChild([=] {
+      ColorScope Color(OS, ShowColors, DeclNameColor);
+      OS << "BoundsFact: " << BF->getVarDecl()->getQualifiedNameAsString();
+      // Visit(BF->getVarDecl());
+      Visit(BF->getBoundsExpr());
```

Comment:
In function `VisitVarDecl`, the declared bounds are being dumped using the function `dumpBoundsAnnotations`. This function visits the bounds expression and also visits the `InteropTypeExpr`. Perhaps `dumpBoundsAnnotations` must be called on line 381?

Reply:
Let me investigate this.

Reply:
A `VarDecl` can have a `InteropTypeExpr` and a `BoundsExpr` e.g. the dump result of `int *g_arr5  : bounds(g_arr5, g_arr5 + 5)` is

```
`-VarDecl 0x55ed451a0248 <line:200:1, col:6> col:6 used g_arr5 'int *'
  |-RangeBoundsExpr 0x55ed451a03f0 <col:16, col:41> 'NULL TYPE'
  | |-ImplicitCastExpr 0x55ed451a03b8 <col:23> 'int *' <LValueToRValue>
  | | `-DeclRefExpr 0x55ed451a02d0 <col:23> 'int *' lvalue Var 0x55ed451a0248 'g_arr5' 'int *'
  | `-BinaryOperator 0x55ed451a0390 <col:31, col:40> 'int *' '+'
  |   |-ImplicitCastExpr 0x55ed451a0358 <col:31> 'int *' <LValueToRValue>
  |   | `-DeclRefExpr 0x55ed451a0300 <col:31> 'int *' lvalue Var 0x55ed451a0248 'g_arr5' 'int *'
  |   `-IntegerLiteral 0x55ed451a0330 <col:40> 'int' 5
  `-InteropTypeExpr 0x55ed451a0448 <<invalid sloc>> '_Array_ptr<int>'
    `-PointerType 0x55ed45177f40 '_Array_ptr<int>'
      `-BuiltinType 0x55ed45135310 'int
```

The dump is done by `dumpBoundsAnnotations`, which prints the bounds and the interop.
The `InteropTypeExpr` is for casting the type of the bound to the declared type (`int` here), and
it's meaningful to dump a faithful `DeclStmt`

Though a where-clause has a `VarDecl`, I think the `Interop` can be omit here because we are 
more interested in the variable name when dumping a bounds in a where-clause.

Reply:
Ok. Sounds reasonable.